### PR TITLE
Allow devices to require a connected display

### PR DESCRIPTION
**Pull request by hughsie**
Fixes https://github.com/fwupd/fwupd/issues/6187

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/linux-drm/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+---
+
+## Introduction
+
+This plugin checks if there are monitors connected to each DRM device. The result will be used to
+inhibit devices that require a monitor to be attached.
+
+## External Interface Access
+
+This plugin requires read access to `/sys/class/drm`.
```
Comment (superm1) [inline plugins/linux-drm/README.md:12]: @valentindavid can you please check if snap rules need to be updated for this new access?

File: plugins/linux-drm/meson.build [hunk 2]
```diff
@@ -0,0 +1,11 @@
+cargs = ['-DG_LOG_DOMAIN="FuPluginLinuxDrm"']
```
Comment (superm1) [inline plugins/linux-drm/meson.build:1]: This needs a check to only load the plugin in Linux.

File: plugins/linux-drm/fu-linux-drm-plugin.c [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	self->idle_id = 0;
+	return FALSE;
+}
+
+static void
+fu_linux_drm_plugin_get_display_state_in_idle(FuLinuxDrmPlugin *self)
+{
+	if (self->idle_id != 0)
+		g_source_remove(self->idle_id);
+	self->idle_id = g_timeout_add_seconds(FU_LINUX_DRM_PLUGIN_REFRESH_DELAY,
```
Comment (superm1) [inline plugins/linux-drm/fu-linux-drm-plugin.c:58]: Is there no uevent to use? 

This makes me wonder if maybe it's better to use a drm ioctl call than parsing the sysfs.

**Review by superm1 — COMMENTED** [review]
Another interesting use for this is any of those mst hubs that only work when a display is connected to them. I know synaptics behaves this way.

File: plugins/linux-drm/README.md (hunk 1 above)
Comment (valentindavid) [reply plugins/linux-drm/README.md:12]: This is provided by interface `opengl`. But I am not sure we should connect to that. Maybe we should just add it to `fwupd` interface. I will bring it up.

File: plugins/linux-drm/fu-linux-drm-plugin.c (hunk 3 above)
Comment (superm1) [reply plugins/linux-drm/fu-linux-drm-plugin.c:58]: There is a uevent: https://dri.freedesktop.org/docs/drm/gpu/drm-uapi.html#sysfs-support.  It's a lot cleaner to register to that IMO.

Comment (hughsie) [reply plugins/linux-drm/fu-linux-drm-plugin.c:58]: Ohh we use that now -- that's the `plugin_class->backend_device_changed` vfunc -- the only reason I added the idle step is at startup if you have 7 devices we'd re-calculate the global state for each device add -- so you'd have, 1, 1+2, 1+2+3, 1+2+3+4 all the way up to 7. The other way (arguably cleaner) is to have a new plugin vfunc of "i've done all the device probing, now recalculate global state" and remove all the timeout stuff.

File: plugins/linux-drm/meson.build (hunk 2 above)
Comment (hughsie) [reply plugins/linux-drm/meson.build:1]: I wasn't sure about the name either -- maybe linux-display?

Comment (hughsie) [reply plugins/linux-drm/meson.build:1]: > This needs a check to only load the plugin in Linux.

Runtime check or build-time? We've not done the other `linux` plugins at build time, but maybe we should...

Comment (superm1) [reply plugins/linux-drm/meson.build:1]: I was thinking build time... I thought we do that already for the other linux ones but looking at it, it's **implicit**.

```
if hsi
```

File: plugins/linux-drm/fu-linux-drm-plugin.c (hunk 3 above)
Comment (superm1) [reply plugins/linux-drm/fu-linux-drm-plugin.c:58]: Ah, I missed that.  Yeah that does sound much cleaner, thanks!

File: plugins/linux-drm/meson.build (hunk 2 above)
Comment (hughsie) [reply plugins/linux-drm/meson.build:1]: Fixed

File: plugins/linux-drm/fu-linux-drm-plugin.c (hunk 3 above)
Comment (hughsie) [reply plugins/linux-drm/fu-linux-drm-plugin.c:58]: https://github.com/fwupd/fwupd/pull/6198

Comment (hughsie) [reply plugins/linux-drm/fu-linux-drm-plugin.c]: Fixed, and rebased.

File: plugins/linux-display/fu-linux-display-plugin.c [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+static gboolean
+fu_linux_display_plugin_plugin_backend_device_added(FuPlugin *plugin,
+						    FuDevice *device,
+						    FuProgress *progress,
+						    GError **error)
+{
+	FuLinuxDisplayPlugin *self = FU_LINUX_DISPLAY_PLUGIN(plugin);
+	if (!fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "drm", error))
+		return FALSE;
+	g_debug("device added: %s", fu_device_get_id(device));
```
Comment (superm1) [inline plugins/linux-display/fu-linux-display-plugin.c:58]: There aren't any IDs specified at the time this runs AFAICT.
```
01:26:53.280 FuEngine             /sys/devices/pci0000:00/0000:00:08.1/0000:63:00.0/drm/card1 added FuUdevDevice:
  Flags:                none
  BackendId:            /sys/devices/pci0000:00/0000:00:08.1/0000:63:00.0/drm/card1
  AcquiesceDelay:       2500
  PossiblePlugin:       linux_display
  Subsystem:            drm
  DeviceFile:           /dev/dri/card1
  SysfsPath:            /sys/devices/pci0000:00/0000:00:08.1/0000:63:00.0/drm/card1

Loading…                 [****************                       ]01:26:53.280 FuEngine             /sys/devices/pci0000:00/0000:00:08.1/0000:63:00.0/drm/card1 added FuUdevDevice:
  Flags:                none
  Guid[quirks]:         129bda59-932a-5210-b863-6f46f0a87584 ← DRM\TYPE_DRM_MINOR
  BackendId:            /sys/devices/pci0000:00/0000:00:08.1/0000:63:00.0/drm/card1
  AcquiesceDelay:       2500
  PossiblePlugin:       linux_display
  Subsystem:            drm
  DeviceFile:           /dev/dri/card1
  SysfsPath:            /sys/devices/pci0000:00/0000:00:08.1/0000:63:00.0/drm/card1

01:26:53.280 FuPlugin             backend_device_added(linux_display)
01:26:53.280 FuPluginLinuxDisplay device added: (null)
```

**Review by superm1 — APPROVED** [review]
Just think the debugging messages need to be improved, otherwise this LGTM.

File: plugins/linux-display/fu-linux-display-plugin.c (hunk 4 above)
Comment (hughsie) [reply plugins/linux-display/fu-linux-display-plugin.c]: Fixed, thanks.

File: plugins/linux-drm/README.md (hunk 1 above)
Comment (valentindavid) [reply plugins/linux-drm/README.md]: @superm1  Does it really need to read that directory, that is to list the symlinks that are there? Or does it need read access to devices of that subsystem? I feel it is porbably needs to be able to read `/sys/devices/**/drm` and `/sys/devices/**/drm/**`

Comment (superm1) [reply plugins/linux-drm/README.md]: I believe you're right

**Pull request merged**